### ISSUE TITLE
关于corr_map = F.softmax(corr_map, dim=-1)的问题

### ISSUE THREAD

**Issue by pascalcpp** [BBBBchan/CorrMatch#27]
感谢您的贡献。我对你的代码中的corr_map = F.softmax(corr_map, dim=-1)这里其实是对行维度进行softmax，但是dict_return['out'] = rearrange(torch.matmul(out_temp, corr_map), 'n c (h w) -> n c h w', h=h_in, w=w_in)，这里获得加权logits是通过列维度得到的，而列维度，即corr_map[0,:,0].sum() != 1。请问这里是否有问题呢，我将改成corr_map = F.softmax(corr_map, dim=1)进行实验后发现性能很差。希望您解答我的疑惑。

**Comment by BBBBchan** [comment]
您好，感谢您对我们工作的关注。这是一个很好的问题，我们在进行pixel propagation时，主要考虑的是变量维度的变化，即用维度为(n, c, hw)的`out_temp`与维度为(n, hw, hw)的`corr_map`相乘，得到的结果仍能保持(n, c, hw)的维度，可以展开为(n, c, h, w)作为一个新的输出分支。
理论上，在pixel propagation时，您在列维度进行softmax操作也是合理的，但这会对后续的region propagation产生影响，这可能是造成性能下降的一个原因。一个可能修改方式可能是同步修改`corr_map_sample = self.sample(corr_map.transpose(1,2).detach(), h_in, w_in)`。
期待您的进一步回复。

**Comment by pascalcpp** [comment]
感谢您的回复。实际上我为了避免其他component的影响，只使用了corr_out来计算损失。而我发现使用行维度即corr_map = F.softmax(corr_map, dim=-1)，相较于corr_map = F.softmax(corr_map, dim=1)会得到更好的结果，这里我只改变了softmax的维度。但是如果使用行维度softmax，会存在corr_map[0,:,0].sum() != 1，这里从直观感觉是存在问题的，因为计算corr_out的加权和中corr_map的不同像素权重不等于1。

**Comment by BBBBchan** [comment]
您好，在我的理解中，您的方式希望能够严格的将逐行的相似度整合到将新的输出分支中。在这种情况下，也许修改代码为`dict_return['out'] = rearrange(torch.matmul(out_temp, corr_map.transpose(1,2)), 'n c (h w) -> n c h w', h=h_in, w=w_in)`可能更加合适？
至于使用corr_map = F.softmax(corr_map, dim=1)得到较差结果的原因，这里给出一个我们的猜想：由于softmax函数的性质，使用`F.softmax(corr_map, dim=1)`时，个别相似度高的像素的相似度（例如该像素本身）会尤为突出，导致不能很好的融合其他同属一类的像素的预测（可以参考correlation map的可视化中属于同类别的像素数量）。

**Comment by pascalcpp** [comment]
非常感谢您的回答，我会认真考虑你的建议。
